clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/ApplicationCachingAgent.java

```diff
@@ -80,10 +119,44 @@ public String getAgentType() {
       Collection<Application> applications) {
     Collection<CacheData> applicationData = new LinkedList<>();
 
+    for (Application application : applications) {
+      Map<String, Object> attributes = convertApplicationToAttributes(application);
+      String applicationKey = Keys.getApplicationKey(application.getName());
```

Comment:
(apologies in advance for the sort-of-handwavy journey this question takes) 
I know right now we're running one of these ApplicationCachingAgents for each ECS account, however each agent is describing _all_ ECS services and divvying them up into their respective applications, so there's some duplication of work in that the diff agents are always overwriting each other's entries. 

It looks like `cacheView.filterIdentifiers()` [supports wildcard matches in the key pattern](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/cache/client/AbstractCacheClient.java#L96-L101), so I'm wondering if it's not possible to index applications by account in addition to provider and namespace, and then retrieve keys/entries that match on provider, namespace and application (but "*" for account), and then flatten the list of services per app before it's returned by the application provider. 

That way, each instance of the Application agent would only be responsible for storing the app/service relationships of a single account (so the individual entries would be smaller, too!), and it would never overwrite data entered by another agent instance. I think it's also valuable to keep track of the application/account relationship b/c it could simplify evicting these in the event the account goes away (like in a dynamic account plugin scenario).

Does the above approach & its advantages makes sense? Perhaps the dynamic account eviction scenario would be taken care of some other way, but I think the smaller key + avoiding overwrites advantages are still compelling.

LMK what you think, or if I missed something that invalidates this.

Reply:
FYI: discussed & tested this separately, and it works but creates similar problems as the one we're trying to solve (i.e., needing to find a whole lot of stuff and then flatten it into one result) so not pursuing it.